src/com/goide/util/GoStringLiteralEscaper.java

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright 2013-2015 Sergey Ignatov, Alexander Zolotov, Mihai Toader, Florin Patan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.goide.util;
+
+import com.goide.psi.GoStringLiteral;
+import com.intellij.openapi.util.Ref;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.LiteralTextEscaper;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * @author stuartcarnie
+ */
+public class GoStringLiteralEscaper extends LiteralTextEscaper<GoStringLiteral> {
+
+  private int[] outSourceOffsets;
+
+  public GoStringLiteralEscaper(@NotNull GoStringLiteral host) {
+    super(host);
+  }
+
+  @Override
+  public boolean decode(@NotNull final TextRange rangeInsideHost, @NotNull final StringBuilder outChars) {
+    TextRange.assertProperRange(rangeInsideHost);
+
+    String subText = rangeInsideHost.substring(myHost.getText());
+
+    if (myHost.getRawString() != null) {
+      outChars.append(subText);
+    }
+    else {
+      Ref<int[]> sourceOffsetsRef = new Ref<int[]>();
+      parseStringCharacters(subText, outChars, sourceOffsetsRef);
+      outSourceOffsets = sourceOffsetsRef.get();
+    }
+
+    return true;
+  }
+
+  @Override
+  public int getOffsetInHost(final int offsetInDecoded, @NotNull final TextRange rangeInsideHost) {
+    TextRange.assertProperRange(rangeInsideHost);
+
+    if (myHost.getRawString() != null) {
+      int offset = offsetInDecoded;
+      offset += rangeInsideHost.getStartOffset();
+      if (offset < rangeInsideHost.getStartOffset()) offset = rangeInsideHost.getStartOffset();
+      if (offset > rangeInsideHost.getEndOffset()) offset = rangeInsideHost.getEndOffset();
+      return offset;
+    }
+
+    int result = offsetInDecoded < outSourceOffsets.length ? outSourceOffsets[offsetInDecoded] : -1;
+    if (result == -1) return -1;
+    return (result <= rangeInsideHost.getLength() ? result : rangeInsideHost.getLength()) + rangeInsideHost.getStartOffset();
+  }
+
+  public static boolean parseStringCharacters(String chars, StringBuilder outChars, Ref<int[]> sourceOffsetsRef) {
```

Comment:
As you mentioned, it was copied from the IntelliJ JSON implementation shared on gitter.im, which is the answer to all your 'why' questions. Easy enough to change visibility and switch to using member variables. It was also changed slightly to suit Go strings.